tt_eager/tensor/tensor.hpp

```diff
@@ -31,35 +31,103 @@ struct Tensor {
+        // This variable is used for on device memory deallocation in async mode, where the main
+        // thread owns all tensors and enqueues a deallocate command for each shard, when a tensor
+        // is implcitly or explictly dellocated.
+        // Call increment when a tensor is default, copy or assignment constructed, since an additonal
+        // object will own a tensor_attr instance.
+        // Call decrement when a tensor is destroyed and the number of owners of the tensor_attr object
+        // decreases.
+        // Record the main thread ref count before pushing to a worker queue (number of owners in main thread).
+        // Update the main thread ref count with the recorded value after the tensor is pushed to the queue(s),
+        // since pushing to the queue requires an extra datacopy in the main thread, that gets balanced by the
+        // worker, howerver the worker cannot modify main_thread_ref_count.
+        void increment_main_thread_ref_count(Device* worker) {
+            if (worker->get_worker_mode() == Device::WorkerQueueMode::ASYNCHRONOUS and worker->in_main_thread()) {
+                main_thread_ref_count++;
+            }
+        }
+
+        void decrement_main_thread_ref_count(Device* worker) {
+            if (worker->get_worker_mode() == Device::WorkerQueueMode::ASYNCHRONOUS and worker->in_main_thread()) {
+                main_thread_ref_count--;
+            }
+        }
+
+        uint32_t record_main_thread_ref_count() {
+            return main_thread_ref_count;
+        }
+
+        void update_main_thread_ref_count(Device* worker, uint32_t ref_count) {
+            if (worker->get_worker_mode() == Device::WorkerQueueMode::ASYNCHRONOUS and worker->in_main_thread()) {
+                main_thread_ref_count = ref_count;
+            }
+        }
     };
 
     // Shared pointer to all attributes associated with this tensor
     // Can be safely passed between threads when the tensor is copied
     std::shared_ptr<TensorAttributes> tensor_attributes;
-
+    // Tensor gets worker queue handle through the device
+    std::vector<Device*> workers = {};
+    bool deallocate_through_destructor = false;
     // ======================================================================================
     //                                  Hi Level APIs
     // ======================================================================================
     Tensor(const Storage storage, const ttnn::Shape shape, DataType dtype, Layout layout);
     Tensor(const Storage storage, const Shape shape, DataType dtype, Layout layout);
 
     // Default constructor to initialize empty tensor
-    Tensor() : tensor_attributes(std::make_shared<TensorAttributes>()) {}
+    Tensor(std::vector<Device*> workers = {}) : tensor_attributes(std::make_shared<TensorAttributes>()), workers(workers) {
+        if (workers.size()) {
+            if (this->workers.at(0)->in_main_thread()) {
```

Comment:
Why only `workers.at(0)`?

Reply:
The logic here is that all workers will be spawned by the same parent/main thread. So if one of them reports that we're in the main thread, they'll all report the same thing.

Multi-device tensor support isn't in yet. As part of adding it, I'll add a sanity check here to ensure that all workers report the same value for `in_main_thread`.